pyccel/codegen/printing/ccode.py

```diff
@@ -1241,6 +1282,9 @@ def get_declare_type(self, expr):
             if expr.is_ndarray or isinstance(expr.class_type, HomogeneousContainerType):
                 if expr.rank > 15:
```

Comment:
Won't this conflict with the rank of `HomogeneousContainerType`? is this relicate from ndarrays still relevant?

Reply:
@EmilyBourne I'm not familiar with this case, could you provide more details?

Reply:
I'm not sure what you mean by "conflict with the rank". Ndarrays are restricted to maximum rank 15 due to restrictions in our implementation. This is not the case for lists but I also would not advise defining a `[[[[[[[[[[[[[[[[]]]]]]]]]]]]]]]]`. That said the code is currently:
```python
if expr.is_ndarray or isinstance(expr.class_type, HomogeneousContainerType):
                if expr.rank > 15:
                    errors.report(UNSUPPORTED_ARRAY_RANK, symbol=expr, severity='fatal')
                if isinstance(expr.class_type, (HomogeneousSetType, HomogeneousListType)):
                    dtype = self.find_in_dtype_registry(expr.class_type)
                    return dtype
                self.add_import(c_imports['ndarrays'])
                dtype = 't_ndarray'
```
So there does not seem to be any gain from grouping NDarrays with other HomogeneousContainerTypes. I think it would make more sense to have:
```python
class_type = expr.class_type
if isinstance(class_type, NDArrayType):
                if expr.rank > 15:
                    errors.report(UNSUPPORTED_ARRAY_RANK, symbol=expr, severity='fatal')
                    return dtype
                self.add_import(c_imports['ndarrays'])
                dtype = 't_ndarray'
elif isinstance(class_type, (HomogeneousSetType, HomogeneousListType)):
                dtype = self.find_in_dtype_registry(expr.class_type)
                return dtype
elif isinstance(class_type, FixedSizeNumericType):
     ....
else:
    errrors.report(...)
```

Reply:
can you check the update in this commit 941703584992ce6d99ca196ee146ef23b29e7065

Reply:
Looks good to me

Reply:
Since `tuples` are implemented using `ndarrays`, this change caused the `tuple` tests to fail. By simply checking `lists` and `sets` in the first part and then handling the rest of the homogeneous containers (such as `ndarrays` and `tuples`) in the second part, the restriction of `rank ≤ 15` is applied only to `ndarrays` and `tuples`.
f3b368b983e387834e4863df0597199befb38f83

Reply:
Ah, good catch! I forgot about tuples. I would rather check `if isinstance(class_type, (NDArrayType, HomogeneousTupleType)):` though. Dicts and cuda arrays will be homogeneous containers and it will be easier to find the places where they are missing support if we have clean errors instead of automatically treating them like ndarrays

Reply:
updated 1134b8e52e17fcc75c1c4123d5820f685ff9c167

Reply:
@jalalium Can we close this conversation? As per the [docs](https://github.com/pyccel/pyccel/blob/devel/developer_docs/review_process.md)
> Please only resolve conversations that you opened. You may think you fixed the problem, but the reviewer may disagree and leaving the discussion open makes it easier for them to verify that they agree with you. If you are reviewing then please close all conversations that you open once the problem is resolved. If you don't this can block the merge.